core/federated/RTI/rti_common.c

```diff
@@ -316,8 +316,15 @@ static void _update_min_delays_upstream(
     // but for most programs, the gain might be negligible since there are relatively few
     // upstream nodes.
     for (int i = 0; i < intermediate->num_upstream; i++) {
-        // Add connection delay to path delay so far.
-        tag_t path_delay = lf_delay_tag(delay_from_intermediate_so_far, intermediate->upstream_delay[i]);
+        // Add connection delay to path delay so far. Because the delay addition is not commutative
+        // but associative, we have to convert the delay to a tag and add to the path delay.
+        tag_t connection_delay = (tag_t) {.time = intermediate->upstream_delay[i], .microstep = 0};
+        if (intermediate->upstream_delay[i] == NEVER) {
+            connection_delay = ZERO_TAG;
+        } else if (intermediate->upstream_delay[i] == 0) {
+            connection_delay.microstep++;
+        }
+        tag_t path_delay = lf_tag_add(connection_delay, delay_from_intermediate_so_far);
```

Comment:
I'm confused about this change.  Isn't this exactly what `lf_delay_tag` is supposed to do?  Perhaps the bug is actually in the implementation of that function?

Reply:
The most important change in this part is the order of operands. In `main`, we calculate `delay_from_intermediate_so_far` + `connection_delay`. However, when a tagged message propagates, the message's tag will be calculated like `intermediate->upstream_delay[i]` + `delay_from_intermediate_so_far`. For example, if we are finding the minimum delay between A and C where the structure looks like A --/1 sec/-->B--/0/-->C and `intermediate` is B, `intermediate->upstream_delay[i]` is `1 sec`, and `delay_from_intermediate_so_far` is `(0, 1)`. The proper minimum delay is `(1 sec, 1) = 1 sec + (0, 1) ≠  (0, 1) + 1 sec`. 

Here, the problem with `lf_delay_tag` is, that it operates `tag_t` + `instant_t` where `connection_delay` is `instant_t` and `delay_from_intermediate_so_far` is `tag_t`. 

Thus, what I did here is, convert `intermediate->upstream_delay[i])` (`instant_t`) to `tag_t` and add `delay_from_intermediate_so_far` to it using `lf_tag_add`. An alternative way to solve this problem would be 
```
tag_t connection_delay = lf_delay_tag(ZERO_TAG, intermediate->upstream_delay[i]);
tag_t path_delay = lf_tag_add(connection_delay, delay_from_intermediate_so_far);
```

Reply:
I see!  Good catch!  This really highlights the value of unit tests.